### ISSUE TITLE
GKE VXLan tunnel is not working

### ISSUE THREAD

**Issue by haiodo** [networkservicemesh/networkservicemesh#886]
## Expected Behavior

VXLAN_TUNNEL should work

## Current Behavior

VXLAN_TUNNEL is not working inside GKE. Working well with VETH mode forced inside Vagrant.

## Steps to Reproduce

Execute any of remote integration tests inside GKE.

## Failure Logs

Dataplane logs for both nodes:

[dp1.log](https://github.com/networkservicemesh/networkservicemesh/files/3022470/dp1.log)
[dp2.log](https://github.com/networkservicemesh/networkservicemesh/files/3022471/dp2.log)

SRC dataplane:
```
root@nsmd-dataplane-gke-dev-default-pool-8cdecbc6-4gw9:~# vppctl show int
              Name               Idx    State  MTU (L3/IP4/IP6/MPLS)     Counter          Count     
host-SRC-1-veth                   3      up          9000/0/0/0     rx packets                   433
                                                                    rx bytes                   18798
                                                                    drops                        433
host-eth0                         1      up          9000/0/0/0     rx packets                    26
                                                                    rx bytes                    1624
                                                                    tx packets                     1
                                                                    tx bytes                      42
                                                                    drops                         25
                                                                    ip4                            3
                                                                    ip6                            7
local0                            0     down          0/0/0/0       
vxlan_tunnel0                     2      up           0/0/0/0       tx packets                   433
                                                                    tx bytes                   34386
```

NSC:
```
kubectl exec -ti nsc-1 ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
3: eth0@if112: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1460 qdisc noqueue state UP 
    link/ether 0a:58:0a:28:00:6d brd ff:ff:ff:ff:ff:ff
    inet 10.40.0.109/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::1866:2fff:fe71:b431/64 scope link 
       valid_lft forever preferred_lft forever
4: nsm@if5: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether ee:03:89:92:26:e4 brd ff:ff:ff:ff:ff:ff
    inet 10.20.1.1/30 brd 10.20.1.3 scope global nsm
       valid_lft forever preferred_lft forever
    inet6 fe80::ec03:89ff:fe92:26e4/64 scope link 
       valid_lft forever preferred_lft forever
```

DST Dataplane:
```
 vppctl show int
              Name               Idx    State  MTU (L3/IP4/IP6/MPLS)     Counter          Count     
host-DST-1-veth                   3      up          9000/0/0/0     rx packets                    23
                                                                    rx bytes                    1730
                                                                    drops                         23
host-eth0                         1      up          9000/0/0/0     rx packets                    24
                                                                    rx bytes                    1568
                                                                    tx packets                     1
                                                                    tx bytes                      42
                                                                    drops                         23
                                                                    ip4                            3
                                                                    ip6                            8
local0                            0     down          0/0/0/0       
vxlan_tunnel0                     2      up           0/0/0/0       tx packets                    23
                                                                    tx bytes                    2558
```

NSE:
```
kubectl exec -ti icmp-responder-nse-1 ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
3: eth0@if43: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1460 qdisc noqueue state UP 
    link/ether 0a:58:0a:28:01:28 brd ff:ff:ff:ff:ff:ff
    inet 10.40.1.40/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::e447:8aff:fefa:a606/64 scope link 
       valid_lft forever preferred_lft forever
4: nsmbnWZ6p2l@if5: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP qlen 1000
    link/ether f2:2f:d1:f4:ef:6c brd ff:ff:ff:ff:ff:ff
    inet 10.20.1.2/30 brd 10.20.1.3 scope global nsmbnWZ6p2l
       valid_lft forever preferred_lft forever
    inet6 fe80::f02f:d1ff:fef4:ef6c/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by haiodo** [comment]
Attaching VETH_MODE enabled Vagrant based logs for local case with working scenario.
[dp_master.log](https://github.com/networkservicemesh/networkservicemesh/files/3022698/dp_master.log)
[dp_worker.log](https://github.com/networkservicemesh/networkservicemesh/files/3022699/dp_worker.log)

**Comment by haiodo** [comment]
Pool request with configuration are here: https://github.com/networkservicemesh/networkservicemesh/pull/867

**Comment by edwarnicke** [comment]
Some quick things to check:

1.  I vaguely remember that 'show int addr' will show us L2 cross connect status.  Worth checking.
2. Try running a [trace](https://wiki.fd.io/view/VPP/Build,_install,_and_test_images#Step_5:_Start_a_packet_trace)... though in your case the interesting inputs are likely to be af-packet-input and vxlan-input rather than dpdk-input.  Also consider increasing the count of packets you trace for from 10 if you don't see what you are looking for.

Other helpful things to try.  You can get access to the VMs GKE is running on.  Try getting some tcpdump captures from them on the HW nic.  Likely its a good idea to restrict them to UDP.  Take a look there and see if you see the VXLAN packets coming from the src and/or getting to the dst.

Basically, we are tracing out the places the packets could get lost:
1. Packets could get lost in the initial src vpp by not being xconed to the vxlan tunnel.
2. Packets could get lost on their way out of src vpp vxlan encaped (unlikely)
3. Packets could get lost on their way out of the src box via the HW NIC
4. Packets could get lost on their way through the network to the dst.
5. Packets could get lost on their way into the dst box
6. Packets could get lost on their way into the dst vpp vxlan 
7.  Packets could get lost cross connecting to the to the dst interface in the dst vpp.

Once you know the break in the chain, it gets much easier to solve the issue :)

**Comment by milanlenco** [comment]
My guess is that this is the infamous checksum-invalid error on the VPP side. In Contiv we had this with both TAPs and VETHs and had to turn off checksum offloading on the pod-side of the VETH/TAP. Later they fixed it for the TAPs so that the offloading could have been re-enabled, but nobody cared about VETHs anymore...
Try turning off the offloading for the VETH in the destination container (i.e. the one with `namespace:<type:FD reference:... >`): https://github.com/ligato/vpp-agent/blob/master/api/models/linux/interfaces/interface.proto#L39-L45
Btw. here we do it in the Contiv: https://github.com/contiv/vpp/blob/master/plugins/ipnet/pod.go#L266-L269

**Comment by edwarnicke** [comment]
[Debugging GKE nodes](https://esc.sh/blog/debugging-gke-nodes/)

**Comment by haiodo** [comment]
Thanks @milanlenco, checked with checksum disabled, not helped, will continue digging.

**Comment by haiodo** [comment]
What we have discovered today:
packets are seems to be dropped on source dataplance side:

In case of drops we have following:

```
vpp# show vxlan tunnel
[0] instance 0 src 10.40.0.13 dst 10.40.1.20 vni 2 fib-idx 0 sw-if-idx 2 encap-dpo-idx 0 
```

```
vpp# show ip fib
ipv4-VRF:0, fib_index:0, flow hash:[src dst sport dport proto ] locks:[src:plugin-hi:2, src:default-route:1, ]
0.0.0.0/0
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:1 buckets:1 uRPF:0 to:[0:0]]
    [0] [@0]: dpo-drop ip4
0.0.0.0/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:2 buckets:1 uRPF:1 to:[0:0]]
    [0] [@0]: dpo-drop ip4
10.40.0.0/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:10 buckets:1 uRPF:9 to:[0:0]]
    [0] [@0]: dpo-drop ip4
10.40.0.0/24
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:9 buckets:1 uRPF:12 to:[0:0]]
    [0] [@4]: ipv4-glean: host-eth0: mtu:9000 ffffffffffff0a580a28000d0806
10.40.0.13/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:12 buckets:1 uRPF:13 to:[0:0]]
    [0] [@2]: dpo-receive: 10.40.0.13 on host-eth0
10.40.0.255/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:11 buckets:1 uRPF:11 to:[0:0]]
    [0] [@0]: dpo-drop ip4
10.40.1.20/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:13 buckets:1 uRPF:0 to:[0:0]]
    [0] [@0]: dpo-drop ip4
224.0.0.0/4
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:4 buckets:1 uRPF:3 to:[0:0]]
    [0] [@0]: dpo-drop ip4
240.0.0.0/4
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:3 buckets:1 uRPF:2 to:[0:0]]
    [0] [@0]: dpo-drop ip4
255.255.255.255/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:5 buckets:1 uRPF:4 to:[0:0]]
    [0] [@0]: dpo-drop ip4

```

And pod info:

```
kubectl get pods -o wide
NAME                                                READY   STATUS    RESTARTS   AGE   IP           NODE                                 NOMINATED NODE
icmp-responder-nse-1                                1/1     Running   0          1h    10.40.1.21   gke-dev-default-pool-785169a3-rvc7   <none>
nsc-1                                               1/1     Running   0          1h    10.40.0.15   gke-dev-default-pool-785169a3-cf9k   <none>
nsmd-dataplane-gke-dev-default-pool-785169a3-cf9k   1/1     Running   0          1h    10.40.0.13   gke-dev-default-pool-785169a3-cf9k   <none>
nsmd-dataplane-gke-dev-default-pool-785169a3-rvc7   1/1     Running   0          1h    10.40.1.20   gke-dev-default-pool-785169a3-rvc7   <none>
nsmgr-gke-dev-default-pool-785169a3-cf9k            3/3     Running   0          1h    10.40.0.14   gke-dev-default-pool-785169a3-cf9k   <none>
nsmgr-gke-dev-default-pool-785169a3-rvc7            3/3     Running   0          1h    10.40.1.19   gke-dev-default-pool-785169a3-rvc7   <none>
```

So for tunnel we have following entry:

```
10.40.1.20/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:13 buckets:1 uRPF:0 to:[0:0]]
    [0] [@0]: dpo-drop ip4
```

**Comment by lobkovilya** [comment]
Vagrant cluster, veth mode, everything works fine:

Tunnel on master dataplane:
```
vpp# show vxlan tunnel
[0] instance 0 src 10.32.0.5 dst 10.44.0.2 vni 2 fib-idx 0 sw-if-idx 2 encap-dpo-idx 1
```

Fib on master dataplane:
```
vpp# show ip fib
ipv4-VRF:0, fib_index:0, flow hash:[src dst sport dport proto ] locks:[src:plugin-hi:2, src:adjacency:1, src:default-route:1, ]
0.0.0.0/0
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:1 buckets:1 uRPF:0 to:[0:0]]
    [0] [@0]: dpo-drop ip4
0.0.0.0/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:2 buckets:1 uRPF:1 to:[0:0]]
    [0] [@0]: dpo-drop ip4
10.32.0.0/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:10 buckets:1 uRPF:9 to:[0:0]]
    [0] [@0]: dpo-drop ip4
10.32.0.0/12
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:9 buckets:1 uRPF:12 to:[0:0]]
    [0] [@4]: ipv4-glean: host-eth0: mtu:9000 ffffffffffff8ac1f5dd3cf00806
10.32.0.5/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:12 buckets:1 uRPF:13 to:[4:424]]
    [0] [@2]: dpo-receive: 10.32.0.5 on host-eth0
10.44.0.2/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:13 buckets:1 uRPF:8 to:[0:0]]
    [0] [@5]: ipv4 via 10.44.0.2 host-eth0: mtu:9000 f2bdb02005a08ac1f5dd3cf00800
10.47.255.255/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:11 buckets:1 uRPF:11 to:[0:0]]
    [0] [@0]: dpo-drop ip4
224.0.0.0/4
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:4 buckets:1 uRPF:3 to:[0:0]]
    [0] [@0]: dpo-drop ip4
240.0.0.0/4
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:3 buckets:1 uRPF:2 to:[0:0]]
    [0] [@0]: dpo-drop ip4
255.255.255.255/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:5 buckets:1 uRPF:4 to:[0:0]]
    [0] [@0]: dpo-drop ip4
```

Pods:
```
NAME                         READY     STATUS    RESTARTS   AGE       IP          NODE
icmp-responder-nse-1         1/1       Running   0          32m       10.44.0.3   kube-worker
nsc-1                        1/1       Running   0          32m       10.32.0.6   kube-master
nsmd-dataplane-kube-master   1/1       Running   0          33m       10.32.0.5   kube-master
nsmd-dataplane-kube-worker   1/1       Running   0          32m       10.44.0.2   kube-worker
nsmgr-kube-master            3/3       Running   0          33m       10.32.0.4   kube-master
nsmgr-kube-worker            3/3       Running   0          32m       10.44.0.1   kube-worker
```

Entry for tunnel:
```
10.44.0.2/32
  unicast-ip4-chain
  [@0]: dpo-load-balance: [proto:ip4 index:13 buckets:1 uRPF:8 to:[0:0]]
    [0] [@5]: ipv4 via 10.44.0.2 host-eth0: mtu:9000 f2bdb02005a08ac1f5dd3cf00800
```

**Comment by edwarnicke** [comment]
@haiodo Some pointers into the vpp code for looking for hints (@artem-belov might also be helpful):

[vxlan4-encap node declaration](https://github.com/FDio/vpp/blob/master/src/vnet/vxlan/encap.c#L474)

[vxlan4-encap node definition](https://github.com/FDio/vpp/blob/master/src/vnet/vxlan/encap.c#L455)

which leads to [vxlan_encap_inline](https://github.com/FDio/vpp/blob/master/src/vnet/vxlan/encap.c#L67)

**Comment by edwarnicke** [comment]
OK... after a brief discussion with a colleague, I strongly suspect our issue is that the mgmt interface is scooping up the ip/subnet of the eth0... but not picking up the routes, and thus vpp has no routes on how to get to the other host.  If we fix that, I suspect things will work :)

**Comment by edwarnicke** [comment]
If I am correct in my theory that its a missing route, you should be able to hand trial a solution using

```vppctl ip route```
( https://docs.fd.io/vpp/19.01/clicmd_src_vnet_ip.html#clicmd_ip_route )
to try setting the default route in vpp by hand to match the one in the underlying kernel ip space.

**Comment by edwarnicke** [comment]
@haiodo Is this fully resolved?

**Comment by haiodo** [comment]
@edwarnicke all is fine now.
